src/Razor/src/Microsoft.AspNetCore.Razor.LanguageServer/Formatting/DefaultRazorFormattingService.cs

```diff
@@ -0,0 +1,287 @@
+        }
+
+        internal static Dictionary<int, IndentationContext> GetLineIndentationMap(RazorSourceDocument source, IReadOnlyList<FormattingSpan> formattingSpans)
+        {
+            var result = new Dictionary<int, IndentationContext>();
+            var total = 0;
+            for (var i = 0; i < source.Lines.Count; i++)
+            {
+                // Get first non-whitespace character position
+                var lineLength = source.Lines.GetLineLength(i);
+                var nonWsChar = 0;
+                for (var j = 0; j < lineLength; j++)
+                {
+                    var ch = source[total + j];
+                    if (!char.IsWhiteSpace(ch) && !ParserHelpers.IsNewLine(ch))
+                    {
+                        nonWsChar = j;
+                        break;
+                    }
+                }
+
+                // position now contains the first non-whitespace character or 0. Get the corresponding FormattingSpan.
+                if (TryGetFormattingSpanIndex(total + nonWsChar, formattingSpans, out var index))
+                {
+                    var span = formattingSpans[index];
+                    result[i] = new IndentationContext
+                    {
+                        Line = i,
+                        IndentationLevel = span.IndentationLevel,
+                        ExistingIndentation = nonWsChar,
+                        FirstSpan = span,
+                    };
+                }
+                else
+                {
+                    // Couldn't find a corresponding FormattingSpan.
+                    result[i] = new IndentationContext
+                    {
+                        Line = i,
+                        IndentationLevel = -1,
+                        ExistingIndentation = nonWsChar,
+                    };
+                }
+
+                total += lineLength;
+            }
+
+            return result;
+        }
+
+        internal static bool TryGetFormattingSpanIndex(int absoluteIndex, IReadOnlyList<FormattingSpan> formattingspans, out int index)
+        {
+            index = -1;
+            for (var i = 0; i < formattingspans.Count; i++)
+            {
+                var formattingspan = formattingspans[i];
+                var span = formattingspan.Span;
+
+                if (span.Start <= absoluteIndex)
+                {
+                    if (span.End >= absoluteIndex)
+                    {
+                        if (span.End == absoluteIndex && span.Length > 0)
+                        {
+                            // We're at an edge.
+                            // Non-marker spans do not own the edges after it
+                            continue;
+                        }
+
+                        index = i;
+                        return true;
+                    }
+                }
+            }
+
+            return false;
+        }
+
+        private async Task<TextEdit[]> FormatProjectedHtmlDocument(
```

Comment:
This and the below methods are not used anywhere yet. I didn't want to include support for these just yet as there are a lot of edge cases to iron out before they are production ready. You will see that in future PRs.